### DISCUSSION TITLE
Complexity Explorer Tutorial

### DISCUSSION THREAD

**Discussion by tpike3** [mesa/mesa#1192] [General]
Team: Its Official! 

We are building a Mesa Tutorial to be hosted by Santa Fe Institute's [Complexity Explorer](https://www.complexityexplorer.org/). 

We will use this discussion to bring it all together based on our many other disparate conversations.

Linked is our [proposal](https://github.com/projectmesa/mesa/files/8309056/Complexity.Explorer.Course.txt) and the [Complexity Explorer guidelines](https://github.com/projectmesa/mesa/files/8309026/Video.Recording.Guidelines.Complexity.Explorer_Mesa.ABM.version.pdf). 

If I was going to summarize the key points of our talk and guidelines, Complexity Explorer is easy to work with and very supportive. We should look to make 12 minute or less videos and keep them direct and on point due to optimize effectiveness for its global audience.    

Initial Timeline: By June 15th send Complexity Explorer first 2 videos. Intro (discussing ABMs) and Part 1 of Mesa tutorial     

Question 1: What ABM are we going to build?
Question 2: What environment should we use?

**Comment by rht** [comment]
Question 1: I'm voting for Sugarscape because it is classic, and has multiple variations.
Question 2: Google Colab seems to be more responsive than Binder.

**Reply by jackiekazil** [reply]
Agree

**Reply by tpike3** [reply]
Sugarscape and colab it is

**Comment by tpike3** [comment]
I have been chewing on this and it seems the intro part set up mesa on colab etc should be pretty easy. 

However for the 12 minute or less intro what do we want to focus on ---

I think the standard why ABMs, a quick overview of Mesa as an open source, everyone welcome project and then overview of the tutorial plan?

**Reply by jackiekazil** [reply]
What do want people to walk away with in 12 minutes? (Think 12 slides or less.)

**Reply by tpike3** [reply]
I would say: 
- ABMs are an emerging scientific tool to explore complex adaptive system based phenomenon. 
- Mesa is a global open source project to advance this tool and all our welcome
- This tutorial will: 
  - Introductory: walk users through building the Sugarscape (traders) models
  - Advanced: how to make contributions
  - Future ones will be produced as needed (i.e. Mesa-Geo)

**Reply by rht** [reply]
We don't have Sugarscape trader model (G1MT in the example yet. We only have the constant growback version.

**Reply by rht** [reply]
I do have the G1MT code on my local machine, but it's not fully tested yet.

**Comment by tpike3** [comment]
I am starting to code out the model and for a June deadline of our initial videos to SFI to get feedback I am thinking:

1 - ABMs as a tool (do at the end)
2 - Overview of the sugarscape traders model (do first and submit as just overview talking material)
3 - part 1 of ??? (do second and submit as coding material)

The million dollar question particularly for 2 is who wants to be the face of Mesa (@jackiekazil) ??).

**Reply by jackiekazil** [reply]
is this part 1, part 2, part 3? or 3 parts to one 12 minute chunk?

**Reply by tpike3** [reply]
3 separate videos which are 12 minutes or less so pat 1, part 2, part 3

**Comment by tpike3** [comment]
I have started the colab "script" for the tutorial. The idea is to have the code for each section that the presenter can just type in and notes. Like writing the idea is to do it iteratively and constantly refine.

Code -----
[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/edgeofchaos42/ComplexityExplorer/blob/main/Trades_in_SugarScape.ipynb)


In the resources file is the JSON, plus some visual adds. Currently they are a picture of sugarscape from mesa interface, a pic with all the numbers and a pic zoomed in on a hill --- the idea is for the visual aids to be discussed in part 2 when introducing the concept of sugarscape.

**Reply by rht** [reply]
Sugarscape trading is usually called Sugarscape 4, instead of 3.

Is it possible to not print the output of the JSON file? It's too crowded currently.

**Reply by rht** [reply]
For reference, this is the GH link, for quicker access: https://github.com/edgeofchaos42/ComplexityExplorer/blob/main/Trades_in_SugarScape.ipynb

**Reply by jackiekazil** [reply]
Is this for the second one?

**Reply by tpike3** [reply]
I just pushed an update that focused on the scripting aspects. 

Let me know if this works for you and makes the process more clear.  The thought is this is our script not what the students see.

**Reply by tpike3** [reply]
@jackiekazil  this is for part 3 -- part 2 will be an overview of sugarscape and the model

**Reply by tpike3** [reply]
I just pushed another update if you want to check out. The script is a loose guide, I rehearsed it and found as long as I have the general idea it is pretty easy to talk through it. Although my typing is pretty error ridden so I am likely not the the best one to record it.

**Reply by rht** [reply]
Using `eval` is dangerous pattern and overkill. I'd store the sugar and spice distribution data as 2 matrices (2 list-of-list). With the matrix, you can quickly do `plt.imshow` in the demo.

**Reply by tpike3** [reply]
Done ... it is much cleaner now .. thanks

**Comment by wang-boyu** [comment]
Would it make sense to include a section on how to integrate GIS data into ABM?

We can probably get some help from Andrew on class materials.

**Reply by rht** [reply]
I think the theme is to use various levels of Sugarscape models, building up features gradually. There is no Sugarscape model that uses GIS data (yet), so it's probably out of scope.

**Reply by tpike3** [reply]
@wang-boyu  The goal is to build multiple modular pieces over time, start with the intro, then the advanced where we talk contributions and at some point build one for GIS, which could be sugarscape or something else. 

Also keeping each block (intro, advanced etc) module as well will allow us to rerecord a section if we break something in the tutorial with an update.

**Reply by wang-boyu** [reply]
Sure! I completely understand.

Besides, Mesa-Geo is not yet ready for tutorials. Hopefully in the future we could have a section on ABM+GIS using Mesa & Mesa-Geo.
